service/controller/v11/resource/endpoint/current.go

```diff
@@ -0,0 +1,76 @@
+package endpoint
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller/context/finalizerskeptcontext"
+	"github.com/giantswarm/operatorkit/controller/context/resourcecanceledcontext"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/kvm-operator/service/controller/v11/key"
+)
+
+func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
+	pod, err := toPod(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "looking for annotations on pod")
+
+	_, serviceName, err := getAnnotations(*pod, IPAnnotation, ServiceAnnotation)
+	if IsMissingAnnotationError(err) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "annotation is missing on pod")
+		resourcecanceledcontext.SetCanceled(ctx)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for pod")
+
+		return nil, nil
+	} else if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	if key.IsPodDeleted(pod) {
+		a := pod.GetAnnotations()
+		if a == nil {
+			return nil, microerror.Mask(missingAnnotationError)
+		}
+		v, ok := a[key.AnnotationPodDrained]
+		if !ok {
+			return nil, microerror.Mask(missingAnnotationError)
+		}
+		if v == "False" {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "cannot finish deletion of pod due to undrained status")
+			resourcecanceledcontext.SetCanceled(ctx)
+			finalizerskeptcontext.SetKept(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+			return nil, nil
+		}
```

Comment:
I think I saw this code chunk several times. WDYT about extracting it to the `key.IsPodDrainded(pod *corev1.Pod) (bool, err)`?

Also for me it's a weird place to cancel reconciliation. My feeling is it should be in `new*Change` functions, then I think you can even avoid cancelling.

Reply:
In the scope of this story I had to cancel a lot of actions to make a lot of different resources behave properly. I wanted to have a common place for this so you know when a resource is canceled it happens in place X by convention. This should make discovering behaviour of resources across more easy and simple. I figured the best place to do this is in the current state area of things. It is the very first place where a resource lifecycle starts. Canceling at the very beginning prevents unnecessary work against external dependencies like the k8s API. There are also resources like the configmap resource which should be canceled in `GetCurrentState` to prevent executing `GetDesiredState` which failed so far otherwise. 

I am fine with having some common functionality to be called. Can you point to the place where code is duplicated in that sense so we can figure our how to improve this? I do not quite follow.

Reply:
I thought

```
func IsPodDraind(pod *corev1.Pod) (bool, error) {
	a := pod.GetAnnotations()
	if a == nil {
		return false, microerror.Mask(missingAnnotationError)
	}
	v, ok := a[key.AnnotationPodDrained]
	if !ok {
		return false, microerror.Mask(missingAnnotationError)
	}
	return v, nil
}

Reply:
Done.